requests_cache/backends/filesystem.py

```diff
@@ -151,3 +166,236 @@ def _get_extension(extension: Optional[str] = None, serializer=None) -> str:
+        block_bytes: int = 1,
+        maximum_file_bytes: Optional[int] = None,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.maximum_cache_bytes = maximum_cache_bytes
+        self.block_bytes = block_bytes
+        self.maximum_file_bytes = (
+            maximum_file_bytes if maximum_file_bytes is not None else maximum_cache_bytes
+        )
+        if self.maximum_file_bytes > self.maximum_cache_bytes:
+            raise ValueError(
+                f'maximum_file_bytes must be smaller or equal to maximum_cache_bytes ({maximum_cache_bytes})'
+            )
+        if self.block_bytes > self.maximum_file_bytes:
+            raise ValueError(
+                f'block_bytes must be smaller than maximum_file_bytes ({maximum_file_bytes})'
+            )
+        if self.block_bytes < 1:
+            raise ValueError(f'block_bytes must be greater than 0, not {block_bytes}')
+        self._oldest_id_checked: int = 0
+        self.id_file = self.cache_dir / 'last.int'
+        self.size_file = self.cache_dir / 'size.int'
+        self.ids_dir = self.cache_dir / 'ids'
+        self.ids_dir.mkdir(exist_ok=True)
+
+    def get_new_file_id(self) -> int:
+        """The id of the last modified file."""
+        with self._lock:
+            if not self.id_file.is_file():
+                self.id_file.write_text('0')
+                return 0
+            new_id = int(self.id_file.read_text()) + 1
+            self.id_file.write_text(str(new_id))
+            return new_id
+
+    @property
+    def newest_id(self) -> int:
+        """The id of the last added file."""
+        try:
+            return int(self.id_file.read_text())
+        except FileNotFoundError:
+            return 0
+
+    @property
+    def total_bytes(self) -> int:
+        """The total size of all the files in the cache."""
+        if not self.size_file.is_file():
+            return 0
+        return int(self.size_file.read_text())
+
+    def _add_to_total_bytes(self, difference_in_bytes: int):
+        """Change the size of the cache."""
+        with self._lock:
+            difference_in_bytes = self.compute_file_size(self.block_bytes, difference_in_bytes)
+            self.size_file.write_text(str(difference_in_bytes + self.total_bytes))
+
+    def _get_raw(self, key: str):
+        """Return the raw content of a key."""
+        content_dir = self.cache_dir / f'{key}{self.extension}'
+        with self._try_io(key):
+            for content_path in content_dir.iterdir():
+                # we can assume there is only one file in the directory
+                return (
+                    content_path.read_bytes()
+                    if self.is_binary
+                    else content_path.read_text(self.encoding)
+                )
+        raise KeyError(f'File for key {key!r} not found.')
+
+    def __getitem__(self, key: str) -> Any:
+        """Get a value from a key."""
+        return self.deserialize(key, self._get_raw(key))
```

Comment:
This is something I am unsure about: deserialization has be put outside of the lock. How is that? Generally safe? Also, I saw that the error is transformed - so, I wonder if this should be adjusted - what is the policy on this one?

Reply:
Deserialization happens fully in-memory and doesn't share any variables with other threads, so there's no need for a lock there.

As for deserialization errors, the original traceback is logged at DEBUG level, and a short message is logged at ERROR level ([here](https://github.com/requests-cache/requests-cache/blob/82c172c4b98fd90c64c1405ced36c025ca53b806/requests_cache/backends/base.py#L360-L362)). Most of the time, a deserialization error just means the response was serialized with an incompatible version of requests-cache, or with a different serializer, and we just discard it and fetch a new response.